network/data_link/at_parser.c

```diff
@@ -0,0 +1,1145 @@
+static inline void at_sscanf(struct at_buff *src, uint8_t *fmt, uint32_t *p1,
+			     uint32_t *p2);
+static inline void at_strcpy(struct at_buff *dest, const struct at_buff *src);
+static inline void at_strcat(struct at_buff *dest, const struct at_buff *src);
+static inline void at_strcatch(struct at_buff *dest, uint8_t ch);
+static inline void at_set(struct at_buff *dest, uint8_t *buff, uint32_t len);
+
+/*
+ * How this function works it represented by this pseoudocode
+ * For each message and its index {
+ * 	if (ch == message[index]) {
+ * 		index++;
+ * 		if (index == strlen(message))
+ * 			return message_by_id
+ * 		}
+ * 	else
+ * 		index = 0;
+ *
+ * 	return -1;
+ * }
+ */
+static inline int32_t match_messages(const struct at_buff *messages,
+				     uint8_t	*indexes,
+				     uint8_t	nb_messages,
+				     uint8_t	ch)
+{
+	int32_t		count;
+	struct at_buff	*msg;
+	uint8_t		*idx;
+
+	for (count = 0; count < nb_messages; count++) {
+		idx = &indexes[count];
+		msg = (struct at_buff *)&messages[count];
+		if (ch == msg->buff[*idx]) {
+			/* Increment matching index if there is a match */
+			(*idx)++;
+			if (*idx == msg->len)
+				/* Return the index of the match message */
+				return count;
+		} else if (ch == msg->buff[0]) {
+			*idx = 1;
+		} else {
+			*idx = 0;
+		}
+	}
+
+	return -1;
+}
+
+/* Check if a +IPD command have been received and updates the desc */
+static bool is_payload_message(struct at_desc *desc, uint8_t ch)
+{
+	static const struct at_buff	at_ipd = {PUI8("\r\n+IPD,"), 7};
+	/* max_ch_search = at_ipd.len + sizeof("0,1024") */
+	static const uint32_t		max_ch_search = 7 + 6;
+	struct	at_buff	aux;
+	uint32_t	id;
+	uint32_t	len;
+	uint32_t	idx_in_result;
+
+	/* Update ipd_idx until at_ipd message is matched */
+	if (desc->ipd_idx < at_ipd.len) {
+		match_messages(&at_ipd, &desc->ipd_idx, 1, ch);
+		return false;
+	}
+	/* Update ipd_idx until ':' is found. Similar to match_messages but
+	 * the message is not fix so we could use that function */
+	if (ch == ':') {
+		/* A well formated +IPD command arrived */
+		id = 0;
+		idx_in_result = desc->result.len - desc->ipd_idx;
+		at_set(&aux, desc->result.buff + idx_in_result, desc->ipd_idx);
+		if (desc->multiple_conections == 1)
+			at_sscanf(&aux, PUI8("\r\n+IPD,%u,%u"), &id, &len);
+		else
+			at_sscanf(&aux, PUI8("\r\n+IPD,%u"), &len, NULL);
+		desc->result.len -= desc->ipd_idx;
+		desc->ipd_idx = 0;
+
+		/* Update the descriptor */
+		desc->current_conn = (int32_t)id;
+		desc->conn[id].to_read = len;
+
+		return true;
+	} else if ((!(ch >= '0' && ch <= '9') && ch != ',') ||
+		   desc->ipd_idx > max_ch_search) {
+		desc->ipd_idx = 0;
+	} else {
+		desc->ipd_idx++;
+	}
+
+	return false;
+}
+
+/* Check if an asynchronous messages was sent by the module and update desc */
+static bool is_async_messages(struct at_desc *desc, uint8_t ch)
+{
+	const static struct at_buff async_msgs[NB_ASYNC_MESSAGES] = {
+		{PUI8("CLOSED\r\n"), 8}
+	};//, //Not used
```

Comment:
I think we can remove commented and unused code.

Reply:
Agree.

Reply:
done

Reply:
done